kernel/cPrimitives.ml

```diff
@@ -188,54 +211,211 @@ let to_string = function
   | Float64ldshiftexp -> "ldshiftexp"
   | Float64next_up    -> "next_up"
   | Float64next_down  -> "next_down"
+  | Arraymake -> "arraymake"
+  | Arrayget -> "arrayget"
+  | Arraydefault -> "arraydefault"
+  | Arrayset -> "arrayset"
+  | Arraycopy -> "arraycopy"
+  | Arrayreroot -> "arrayreroot"
+  | Arraylength -> "arraylength"
+
+type const =
+  | Arraymaxlength
 
-type prim_type =
-  | PT_int63
-  | PT_float64
+let const_to_string = function
+  | Arraymaxlength -> "arraymaxlength"
+
+let const_of_string = function
+  | "array_max_length" -> Arraymaxlength
+  | _ -> raise Not_found
 
-type 'a prim_ind =
+let const_univs = function
+  | Arraymaxlength -> AUContext.empty
+
+type 'a prim_type =
+  | PT_int63 : unit prim_type
+  | PT_float64 : unit prim_type
+  | PT_array : (Instance.t * ind_or_type) prim_type
+
+and 'a prim_ind =
   | PIT_bool : unit prim_ind
-  | PIT_carry : prim_type prim_ind
-  | PIT_pair : (prim_type * prim_type) prim_ind
+  | PIT_carry : ind_or_type prim_ind
+  | PIT_pair : (ind_or_type * ind_or_type) prim_ind
   | PIT_cmp : unit prim_ind
   | PIT_f_cmp : unit prim_ind
   | PIT_f_class : unit prim_ind
 
-type prim_ind_ex = PIE : 'a prim_ind -> prim_ind_ex
-
-type ind_or_type =
+and ind_or_type =
   | PITT_ind : 'a prim_ind * 'a -> ind_or_type
-  | PITT_type : prim_type -> ind_or_type
+  | PITT_type : 'a prim_type * 'a -> ind_or_type
+  | PITT_param : int -> ind_or_type (* DeBruijn index referring to prenex type quantifiers *)
+
+let one_univ =
+  AUContext.make Names.[|Name (Id.of_string "u")|] Constraint.empty
+
+let typ_univs (type a) (t : a prim_type) = match t with
+  | PT_int63 -> AUContext.empty
+  | PT_float64 -> AUContext.empty
+  | PT_array -> one_univ
+
+type prim_type_ex = PTE : 'a prim_type -> prim_type_ex
+
+type prim_ind_ex = PIE : 'a prim_ind -> prim_ind_ex
 
 let types =
-  let int_ty = PITT_type PT_int63 in
-  let float_ty = PITT_type PT_float64 in
+  let int_ty = PITT_type (PT_int63, ()) in
+  let float_ty = PITT_type (PT_float64, ()) in
+  let array_param =
+    PITT_type
+      (PT_array,
+       (Instance.of_array [|Level.var 0|],
+        PITT_param 1))
+  in
   function
-  | Int63head0 | Int63tail0 -> [int_ty; int_ty]
+  | Int63head0 | Int63tail0 -> [[int_ty]; [int_ty]]
```

Comment:
Why put all the types into length-1 arrays? What are the semantics of length-2 arrays? Is there a use case somewhere?